### DISCUSSION TITLE
Precompiled DLL libraries, for 3d Alpha Wrapping

### DISCUSSION THREAD

**Discussion by mariuszhermansdorfer** [CGAL/cgal#6848] [Q&A] [answered]
## Issue Details

Would it be possible to include precompiled DLL libraries for CGAL releases?

I'm working with c# and don't have any experience with building c++ projects from source. Still, I'd love to be able to call some CGAL functions from my c# code for which I only need a DLL.

## Environment

* Operating system (Windows 64 bits):

**Comment by lrineau** [comment]
Hi Marius. CGAL is a header-only library of C++ templates, and it cannot easily be built into DLL libraries: for that you would need first to create `.cpp` containing the instantiations of the templates, and the API, you need.

What you ask for is actually similar to what was done in https://github.com/CGAL/cgal-swig-bindings, to provide bindings for Java and Python. You actually for a sort of bindings for C++-without-templates. Maybe the DLLs creates by the build of https://github.com/CGAL/cgal-swig-bindings could actually correspond to what you want. I cannot really say more about the subject for now.

**Reply by mariuszhermansdorfer** [reply]
@lrineau, thanks for your prompt response!

As you can probably tell, I'm not at all familiar with how c++ works. In c# there is a concept of `type marshalling`, which allows us to go between managed and unmanaged code. 
https://docs.microsoft.com/en-us/dotnet/standard/native-interop/type-marshalling

For this to work one needs to import a DLL library:
`[DllImport("somenativelibrary.dll")]`

If I understand you correctly, it is not a straightforward process to compile one CGAL DLL for say Windows 64 bit?

**Reply by lrineau** [reply]
No, that is not straightforward. What parts of CGAL are your interested in?

**Reply by mariuszhermansdorfer** [reply]
Specifically, I'd like to have access to 3d Alpha Wrapping:
https://www.cgal.org/2022/05/18/alpha_wrap/

**Reply by lrineau** [reply]
CGAL 3d Alpha Wrapping is a hot topic, since its publication.

Marius, for the rest of the discussion, I want to call colleges of mine:
  - Mael Rouxel-Labbé (@MaelRL), who is one of the co-author of the 3d Alpha Wrapping,
  - Sébastien Loriot (@sloriot), who is mostly the main author of the CGAL Swig bindings.

Mael, Sébastien, as far as I know, the alpha-wrap is not in the CGAL Swig bindings. How difficult could it be to add it? The [reference manual](https://doc.cgal.org/5.5/Alpha_wrap_3/group__PkgAlphaWrap3Ref.html) only mention one function and thus that looks easy.

Sébastien, do you know how difficult it would be to have C# as one of the output language of our bindings?

**Reply by lrineau** [reply]
@mariuszhermansdorfer Actually, as the 3d Alpha Wrapping is only one function template, it might be rather easy to compile one DLL with a good API, with just that feature.

**Reply by mariuszhermansdorfer** [reply]
It would be awesome if we could get it to work! And thanks for being so open for suggestions. I appreciate it a lot!

**Reply by mariuszhermansdorfer** [reply]
Hey @lrineau, do you have any examples on how to approach this problem? I'm willing to learn (and contribute to the repo) but would need to have some guidance on where to start.

**Comment by sloriot** [comment, marked as answer]
See https://github.com/CGAL/cgal-swig-bindings/commit/ca5f0c4e6cb191a94658559bd2e55451bf7ae1db that is adding it. Tested locally in java and python.
